### PR TITLE
Added Operator support for Partitioned HashJoin.

### PR DESCRIPTION
**Pull request by zuyu**
Assigned to @hbdeshmukh or @jianqiao.

This PR allows `HashJoin` to accept partitioned info. If no partition, `num_partitions` is `1`.

### CODE REVIEW

File: storage/StorageBlockInfo.hpp [hunk 1]
```diff
@@ -49,6 +50,8 @@ static constexpr int kBlockIdDomainLengthInDigits = std::numeric_limits<block_id
 static constexpr int kBlockIdCounterLengthInDigits = std::numeric_limits<block_id_counter>::digits10;
 static constexpr block_id_domain kMaxDomain = static_cast<block_id_domain>(0xFFFF);
 
+typedef std::vector<block_id> blocks_in_partition;
```
Comment (jianqiao) [inline storage/StorageBlockInfo.hpp]: `block_in_partition` is not a primitive type. Perhaps use `BlockInPartition`?

File: relational_operators/HashJoinOperator.cpp [hunk 2]
```diff
@@ -207,49 +207,43 @@ bool HashJoinOperator::getAllNonOuterJoinWorkOrders(
...
-                residual_predicate,
-                selection,
-                hash_table,
-                output_destination,
-                storage_manager,
-                CreateLIPFilterAdaptiveProberHelper(lip_deployment_index_, query_context)),
-            op_index_);
-        ++num_workorders_generated_;
-      }  // end while
+      for (std::size_t part_id = 0; part_id < num_partitions_; ++part_id) {
```
Comment (jianqiao) [inline relational_operators/HashJoinOperator.cpp]: How many partitions would we expect to have in typical situations?

Comment (zuyu) [inline relational_operators/HashJoinOperator.cpp]: It depends on the input relation and the optimizer. The former may have given partitions (i.e., 4), and the latter may add an explicit partition phase for a relation w/o partitions.

I do so to minimize the code changes without hurting the performance much.

Comment (jianqiao) [inline relational_operators/HashJoinOperator.cpp]: I see. It should be good.

File: storage/StorageBlockInfo.hpp (hunk 1 above)
Comment (zuyu) [inline storage/StorageBlockInfo.hpp]: Done.

**Comment by jianqiao** [comment]
LGTM. Merging.

**Pull request merged**